### PR TITLE
Configure enabled languages

### PR DESCRIPTION
**Pull request by m5r**
# Description

medic/cht-core#6281

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:6281-configure-disabled-languages/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:6281-configure-disabled-languages/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:6281-configure-disabled-languages/docker-compose/cht-couchdb-clustered.yml)
 
# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: api/src/migrations/extract-enabled-locales.js
Comment (garethbowen) [inline api/src/migrations/extract-enabled-locales.js:47]: As [mentioned in this comment](https://github.com/medic/cht-core/issues/6281#issuecomment-1480622151) I'm not convinced that a migration will work here.

An alternative approach would be to change all the code that...
1. Looks in the settings for `enabledLocales` and if it's found, use it
2. Falls back to the current code when it's not found.

Another option would be to find all languages that are enabled in settings _or_ in the translation doc but that makes it harder to deprecate and remove the old way.

Comment (garethbowen) [reply api/src/migrations/extract-enabled-locales.js:47]: > a first migration that sets the enabledLocales in settings that we release now~ish -- with the fallback logic

I don't think the migration adds anything. What I expect will happen is...

1. During the upgrade the migration will run and update the `settings` doc.
2. Now anyone looking for enabled locales will not need the fallback logic - it'll just work.
3. Soon after the app dev uploads the settings from CI or CLI and overwrites the `settings` doc which removes the `enabledLocales` property.
4. Now we're back to where we started and the fallback logic will run every time

I think it's less risky to just write the fallback logic and not bother with the migration.

Comment (m5r) [reply api/src/migrations/extract-enabled-locales.js:47]: Fair point, that would render the migration useless...

I'll move forward with just the fallback logic and open another issue to deprecate the old way for 5.x

Comment (m5r) [reply api/src/migrations/extract-enabled-locales.js]: Opened up the follow up issue medic/cht-core#8157

File: admin/src/js/controllers/edit-user.js
Comment (garethbowen) [inline admin/src/js/controllers/edit-user.js:40]: Nice catch!

File: api/src/translations.js
Comment (garethbowen) [inline api/src/translations.js:106]: `docs` is a bit generic... What about `translations` or `translationDocs`?

File: admin/src/js/controllers/display-languages.js
Comment (garethbowen) [inline admin/src/js/controllers/display-languages.js:24]: TODO... settings structure - right now it's an array which means we can never extend it with any other info, eg: https://github.com/medic/cht-core/issues/8158

**Review by garethbowen — CHANGES_REQUESTED** [review]
Great work, thanks!

A few comments inline, and I have some more general comments...

1. We should provide default `enabledLocales` configuration, so that new projects and demos work out of the box. At the moment it should probably be just English (because we're missing translations in other languages) but hopefully we can build out a small set of fully supported languages and demos will be slick! Just double check that this doesn't update existing settings when upgrading - I'm 99% sure it doesn't but worth a look.
2. You've set `enabledLocales` as an array, which is fine for this work, but if we ever want to store more information about locales then we'll need a separate settings property which is messy (for example, #8158 ). To be future proof it would be better to have it as an object with an `enabled` boolean property. If you agree so far, then we may need to change the name of the property to be more generic (eg: `locales`) but would have to be careful not to conflict with the old property. If we do decide to call it `locales` then it may be backwards compatible if we check for the `enabled` property, and default to `true`.

This is quite complex, so reach out if anything needs a better explanation!

**Comment by m5r** [comment]
@garethbowen 
> To be future proof it would be better to have it as an object with an enabled boolean property.	

That's a great suggestion and considering the issue you linked to, it would be best to anticipate this change now.

> If we do decide to call it locales then it may be backwards compatible if we check for the enabled property, and default to true.

The issue I see here is that we would end up never using the fallback code branch because I haven't seen a deployment without the `locales` property in their settings despite it being deprecated for a while. In this situation, I suggest we either choose a different property name or we introduce the migration now.  
If we decide to introduce the migration now and an app dev decides to upload the settings from CI or CLI and overwrites the `settings` doc, it wouldn't be an issue **if** `settings.locales` only lists languages supported by the project.

**Comment by garethbowen** [comment]
> I suggest we either choose a different property name or we introduce the migration now.

I agree. While I think `locales` is a really good name, having the old deprecated config and the new config with the same name will only lead to confusion and complicated bugs. I think we just go for a new name so we don't get any collisions and it's clear what's going on. Maybe `languages`? Maybe you can come up with something better...

**Review by jkuester — CHANGES_REQUESTED** [review]
Couple things in no particular order:

1. Don't forget tests!
2. Just now reading the latest dicussion out on the PR, so some of my in-line suggestions are a bit out-dated (assuming `enabledLocales` is not going to be an array....)
3. Also, I agree that re-using `locales` seems like a bad idea.  `langauges` is an okay name, but I would suggest `translations` (unless I missed somthing and that is already taken....)
4.  I get that we cannot have a server-side migration to move the config out of the messages docs and into the settings.  However, the current both/and approach seems prone to confusing behavior (see below) and it forces every place that wants to check the enabled translations (admin/api/webapp) to read _both_ settings and message docs.  That is a bunch of repeated steps.
    - The confusing scenario goes like this:
        - Running old code, admin enables EN and ES
        - CHT upgraded to new version
        - Admin sets `enabledLocales = ['EN', 'ES']` in app settings
        - Admin uses App Management UI to _disable_ `ES` - this updates the settings doc to have `enabledLocales = ['EN']`
        - Admin accidentally pushes settings without `enabledLocales` - this will cause `EN` _and_ `ES` to be enabled since that is what was still stored in the docs `enabled` field.
    - My suggestion here is, at this point, to avoid referencing the `app_settings` at all when _reading_ the enabled translations
        - When reading the translations, all the places (api/webapp/admin) will just read the `medic-client/doc_by_type` view.
        - In a change listener for api or sentinel (not sure which one is a better place) we can listen for `settings` changes and add logic to update the `enabled` field on the message docs whenever the contents of `enabledLocales` does not match what is set on the messages docs 
        - In admin, when enabling/disabling the translation, it should update the `settings` doc if the `enabledLocales` structure already exists else it should update the `enabled` field on the message docs.

File: admin/src/js/controllers/edit-user.js
Comment (garethbowen) [reply admin/src/js/controllers/edit-user.js:40]: I think this used to be used back when you could change another user's locale through the admin app. Back then we had a dropdown of `enabledLocales` so you could pick one. When the ability to change another user's locale was removed we must have forgotten to remove this, so we still got the data, just never used it.

**Comment by m5r** [comment]
@jkuester thanks a lot for your input! I agree that the current approach comes with some overhead but it allows deployments to upgrade without having to change their settings yet, they decide when they want to opt-in this new way of configuring their enabled languages.  
Each project has its own repo tracking `app_settings` changes so once they decide to enable this setting, it's explicitly defined in the file and won't go anywhere unless manually deleted.  
On a side note, we planned to streamline the logic and get rid of the overhead in medic/cht-core#8157

Potentially overriding enabled languages set from the UI is a real risk but I wonder how frequently that happens (vs configuring languages once at the beginning of the project and forgetting about it) and if it's worth the extra complexity? cc @garethbowen

**Comment by m5r** [comment]
I initially went with `translations` but this property was already used in the past and there is the `extract-translations` migration that prevents us from reusing that property so I went with `languages`

**Comment by garethbowen** [comment]
> Potentially overriding enabled languages set from the UI is a real risk but I wonder how frequently that happens (vs configuring languages once at the beginning of the project and forgetting about it) and if it's worth the extra complexity?

Basically everything in the app management app will be overwritten if the project then uses cht-conf so it's acceptable. Essentially you should use one or the other but not both.

**Review by jkuester — CHANGES_REQUESTED** [review]
Still seems really unfortunate that you can have conflicting `enabled` values set in the `messages-xx` doc vs the `settings` doc, but I suppose that is the price of shipping this feature now and not later as a breaking change....

Everything works well! Just need tests!  (Would be nice to have a dedicated e2e test that enables/disables languages via settings without touching the `messages-xx` docs.)

**Review by jkuester — APPROVED** [review]
Tres bien!

File: tests/e2e/default/translations/enabled-languages.wdio-spec.js
Comment (m5r) [reply tests/e2e/default/translations/enabled-languages.wdio-spec.js:100]: I thought so too but we can't because it's meant to log the user out from the webapp, not the admin app

File: tests/e2e/default/translations/enabled-languages.wdio-spec.js
Comment (m5r) [reply tests/e2e/default/translations/enabled-languages.wdio-spec.js:24]: I was seeing some runs failing because the login page didn't have the expected languages where this bit of code helped stabilize the test but I can't reproduce it anymore. I'm getting rid of it!

File: tests/e2e/default/translations/enabled-languages.wdio-spec.js
Comment (jkuester) [reply tests/e2e/default/translations/enabled-languages.wdio-spec.js:78]: Ah, yes!  :facepalm: Not a big deal, but this bit of functionality could also go in `admin.wdio.page.js` since it is specific to the admin console....

Comment (m5r) [reply tests/e2e/default/translations/enabled-languages.wdio-spec.js]: Ah, nice catch!

**Review by tatilepizs — APPROVED** [review]
Thank you Mokhtar for adding and updating the e2e tests 🎉

File: tests/page-objects/default/admin/admin.wdio.page.js
Comment (m5r) [reply tests/page-objects/default/admin/admin.wdio.page.js]: Ooh thanks for catching this!

**Comment by m5r** [comment]
@garethbowen looks like I can't merge until you approve the changes :sweat_smile:

File: admin/tests/unit/controllers/display-languages.js
Comment (garethbowen) [inline admin/tests/unit/controllers/display-languages.js:164]: This test would be a lot cleaner if you changed `disableLanguage` to return the promise. That way instead of setting a timeout and hoping everything happens in the next tick you could just await the response.

**Review by garethbowen — CHANGES_REQUESTED** [review]
Awesome work. Sorry for coming late to the party.

Mostly these suggestions are nitpicks but there's one bug that we should fix.

File: admin/src/js/controllers/display-languages.js
Comment (m5r) [reply admin/src/js/controllers/display-languages.js:71]: Nice catch!

File: admin/tests/unit/controllers/display-languages.js
Comment (m5r) [reply admin/tests/unit/controllers/display-languages.js]: That's much cleaner indeed, thanks!

**Review by garethbowen — APPROVED** [review]
Great work on those tests! They look so good now.

**Pull request merged**
